include/prepared_statement.h

```diff
@@ -0,0 +1,43 @@
+#include <inttypes.h>
+
+/* The format that our prepared statements are called on the Postgres server */
+#define PREPARED_STMT_NAME_FORMAT "PS_%" PRIu64
+/* "PS_" + 20 digits for unsigned 64 bit int + \0 */
+#define MAX_SERVER_PREPARED_STMT_NAME 24
+
+
+/* Structure to store text of prepared query */
+typedef struct PgPreparedStatement {
+	UT_hash_handle	hh;
+	uint64_t	query_id;
+	uint32_t	use_count;
+	uint32_t	query_len;
+	uint32_t	param_len;
+	uint8_t		stmt_name_len;
+	char		stmt_name[MAX_SERVER_PREPARED_STMT_NAME];
+	char		query_with_params[]; /* varying length */
+} PgPreparedStatement;
+
+/* Client session prepared statements */
+typedef struct PgClientPreparedStatement
+{
+	UT_hash_handle	hh;
+	PgPreparedStatement* ps;
+	char 			stmt_name[]; /* varying size */
+} PgClientPreparedStatement;
+
+/* Prepared statements in Postgres backends */
+typedef struct PgServerPreparedStatement
+{
+	uint64_t query_id;
+	UT_hash_handle hh;
+	PgPreparedStatement* ps;
+} PgServerPreparedStatement;
+
+#define is_prepared_statements_enabled(pool) \
+	(pool_pool_mode(pool) == POOL_TX && cf_prepared_statement_cache_size != 0)
```

Comment:
Any reason we cannot allow this in prepared statements in `statement` pooling? I think it should be fine.

Reply:
I do not see any problems with using prepared statements for statement pooling.
But I always wonder why anybody needs statement pooling at all?
We can not rely on any properties of transaction: isolation, atomicity,... if different statements of the same transaction are executed by different backends. So it actually means that we can not use transactions and each statement is executed as separate transaction. But in this case there is no difference between transaction and statement modes, isn't it?
It can be some optimisation allowing pooler not to keep track of transaction boundaries. But we in any case have to parse PQ commands. So it is not clear if we are saving something here by not tracking transaction begin/end.

Reply:
Yeah, I also don't really know if there's really any usecase for statement pooling. It might just be there because it was easy to implement (it has existed since the first version of pgbouncer). Deprecating/removing it seems like a separate discussion. For now I added support for statement pooling to this PR.